firedrake/slate/slac/compiler.py

```diff
@@ -209,32 +207,47 @@ def generate_kernel_ast(builder, statements, declared_temps):
     return kinfo
 
 
-def auxiliary_temporaries(builder, declared_temps):
+def auxiliary_expressions(builder, declared_temps):
     """Generates statements for assigning auxiliary temporaries
-    for nodes in an expression with "high" reference count.
-    Expressions which require additional temporaries are provided
-    by the :class:`LocalKernelBuilder`.
+    and declaring factorizations for local matrix inverses
+    (if the matrix is larger than 4 x 4).
 
     :arg builder: The :class:`LocalKernelBuilder` containing
-                  all relevant expression information.
+        all relevant expression information.
     :arg declared_temps: A `dict` containing all previously
-                         declared temporaries. This dictionary
-                         is updated as auxiliary expressions
-                         are assigned temporaries.
+        declared temporaries. This dictionary is updated as
+        auxiliary expressions are assigned temporaries.
     """
-    statements = [ast.FlatBlock("/* Auxiliary temporaries */\n")]
-    results = [ast.FlatBlock("/* Assign auxiliary temps */\n")]
-    for exp in builder.aux_exprs:
+
+    # These are either already declared terminals or expressions
+    # which do not require an extra temporary/expression
+    terminals = (slate.Tensor, slate.AssembledVector,
+                 slate.Negative, slate.Transpose)
+    statements = []
+
+    sorted_exprs = [exp for exp in topological_sort(builder.expression_dag)
+                    if (builder.ref_counter[exp] > 1 and not isinstance(exp, terminals))
```

Comment:
Parenthesise like `if ((builder.ref_counter[exp] > 1 and not ...) or isinstance(exp, slate.Factorization))` to make it clearer?